synapse/lib/kv.py

```diff
@@ -0,0 +1,466 @@
+
+        Args:
+            vals (list): A list of msgpack values to add to the set.
+
+        Returns:
+            None
+        '''
+        dups = [(self.iden, s_msgpack.en(v)) for v in vals]
+
+        self.vals.update(vals)
+        self.stor.addKvDups(dups)
+
+class KvStor(s_eventbus.EventBus):
+    '''
+    The KvStor uses an LMDB key-value database to implement
+    persistence and indexing for primitive data structures.
+    '''
+
+    def __init__(self, path):
+        s_eventbus.EventBus.__init__(self)
+        self.lenv = lmdb.open(path, writemap=True, max_dbs=16)
+
+        self.oob = self.lenv.open_db(b'oob')
+        self.dups = self.lenv.open_db(b'dups', dupsort=True)
+
+        self.onfini(self._onKvFini)
+
+    def _onKvFini(self):
+        self.lenv.sync()
+        self.lenv.close()
+
+    @functools.lru_cache(maxsize=1024)
+    def genKvAlias(self, name):
+        '''
+        Resolve or create a new object alias by name.
+
+        Args:
+            name (str): String to create or resolve an alias for.
+
+        Notes:
+            The iden returned as an alias is randomly generated the first time
+            that ``genKvAlias`` is called for a given name.
+
+        Returns:
+            bytes: The iden for the name
+        '''
+        lkey = b'alias:' + name.encode('utf8')
+
+        with self.lenv.begin(db=self.oob, write=True) as xact:
+
+            iden = xact.get(lkey)
+            if iden is None:
+                iden = os.urandom(16)
+                xact.put(lkey, iden)
+
+            return iden
+
+    def getKvSet(self, name):
+        '''
+        Create or retrieve a KvSet by name from the KvStor.
+
+        Args:
+            name (str): The name of the KvList.
+
+        Returns:
+            KvSet: The KvList helper instance.
+        '''
+        iden = self.genKvAlias(name)
+        return KvSet(self, iden)
+
+    def getKvDict(self, name):
+        '''
+        Create or retrieve a KvDict by name from the KvStor.
+
+        Args:
+            name (str): The name of the KvDict.
+
+        Returns:
+            KvDict: The KvDict helper instance.
+        '''
+        iden = self.genKvAlias(name)
+        return KvDict(self, iden)
+
+    def getKvLook(self, name):
+        '''
+        Create or retrieve a KvLook by name from the KvStor.
+
+        Args:
+            name (str): The name of the KvLook.
+
+        Returns:
+            KvLook: The KvLook helper instance.
+        '''
+        iden = self.genKvAlias(name)
+        return KvLook(self, iden)
+
+    def iterKvDups(self, lkey):
+        '''
+        Yield lkey, lval tuples for the given dup key.
+
+        Args:
+            lkey (bytes): The kv key.
+
+        Yields:
+            bytes: The value of the dups for a given key.
+        '''
+        with self.lenv.begin(db=self.dups) as xact:
+
+            with xact.cursor(db=self.dups) as curs:
+
+                if not curs.set_key(lkey):
+                    return
```

Comment:
This is the only untested line. I'm uncertain how to force ``set_key`` to return false.  @jnwatson  thoughts?

Reply:
Looks like it will return false if the key is not found. https://lmdb.readthedocs.io/en/release/#lmdb.Cursor.set_key

PR: https://github.com/vEpiphyte/synapse/pull/5 (feel free to merge my PR or delete it and make the change yourself)

Reply:
Merged that PR

Reply:
thanks!